ethereum/datastructures/src/main/java/tech/pegasys/teku/datastructures/blocks/SignedBeaconBlock.java

```diff
@@ -13,88 +13,102 @@
-import com.google.common.base.Suppliers;
-import java.util.ArrayList;
-import java.util.List;
-import java.util.Objects;
 import java.util.Optional;
-import java.util.function.Supplier;
-import jdk.jfr.Label;
-import org.apache.tuweni.bytes.Bytes;
 import org.apache.tuweni.bytes.Bytes32;
 import tech.pegasys.teku.bls.BLSSignature;
-import tech.pegasys.teku.datastructures.util.HashTreeUtil;
-import tech.pegasys.teku.datastructures.util.HashTreeUtil.SSZTypes;
 import tech.pegasys.teku.datastructures.util.Merkleizable;
-import tech.pegasys.teku.datastructures.util.SimpleOffsetSerializer;
+import tech.pegasys.teku.datastructures.util.SpecDependent;
 import tech.pegasys.teku.infrastructure.unsigned.UInt64;
 import tech.pegasys.teku.ssz.SSZTypes.SSZContainer;
+import tech.pegasys.teku.ssz.backing.VectorViewRead;
+import tech.pegasys.teku.ssz.backing.containers.Container2;
+import tech.pegasys.teku.ssz.backing.containers.ContainerType2;
+import tech.pegasys.teku.ssz.backing.tree.TreeNode;
+import tech.pegasys.teku.ssz.backing.type.ComplexViewTypes;
+import tech.pegasys.teku.ssz.backing.view.BasicViews.ByteView;
+import tech.pegasys.teku.ssz.backing.view.ViewUtils;
 import tech.pegasys.teku.ssz.sos.SimpleOffsetSerializable;
+import tech.pegasys.teku.ssz.sos.SszTypeDescriptor;
 
 public class SignedBeaconBlock
+    extends Container2<SignedBeaconBlock, BeaconBlock, VectorViewRead<ByteView>>
     implements BeaconBlockSummary, Merkleizable, SimpleOffsetSerializable, SSZContainer {
 
-  private final BeaconBlock message;
-  private final BLSSignature signature;
+  public static class SignedBeaconBlockType
+      extends ContainerType2<SignedBeaconBlock, BeaconBlock, VectorViewRead<ByteView>> {
 
-  @Label("sos-ignore")
-  private final Supplier<Bytes32> hashTreeRoot = Suppliers.memoize(this::calculateRoot);
+    public SignedBeaconBlockType() {
+      super(
+          "SignedBeaconBlock",
+          namedType("message", BeaconBlock.TYPE.get()),
+          namedType("signature", ComplexViewTypes.BYTES_96_TYPE));
+    }
 
-  public SignedBeaconBlock(final BeaconBlock message, final BLSSignature signature) {
-    this.message = message;
-    this.signature = signature;
+    @Override
+    public SignedBeaconBlock createFromBackingNode(TreeNode node) {
+      return new SignedBeaconBlock(this, node);
+    }
   }
 
-  public BeaconBlock getMessage() {
-    return message;
+  @SszTypeDescriptor
+  public static SignedBeaconBlockType getSszType() {
+    return TYPE.get();
   }
 
-  public BLSSignature getSignature() {
-    return signature;
+  public static final SpecDependent<SignedBeaconBlockType> TYPE =
+      SpecDependent.of(SignedBeaconBlockType::new);
+
+  private BLSSignature signatureCache;
+
+  private SignedBeaconBlock(SignedBeaconBlockType type, TreeNode backingNode) {
+    super(type, backingNode);
   }
 
-  @Override
-  public int getSSZFieldCount() {
-    return message.getSSZFieldCount() + signature.getSSZFieldCount();
+  @Deprecated
```

Comment:
Yes, good question! 
From on side: `SignedBeaconBlockType` depends on `BeaconBlockType` which depends on `BeaconBlockBodyType` which in turn depends on `Spec`. `SignedBeaconBlock` instance should be bounded to its type on construction and we should basically derive `SignedBeaconBlockType` from the `Spec` and pass it to `SignedBeaconBlock` constructor. Else we are ending up using a static `SignedBeaconBlockType` instance which we are trying to eliminate. 

From the other side: we have a `BeaconBlock` and may get it's type and create `SignedBeaconBlockType` dynamically. However `ViewType` classes were originally designed to be more or less long living since they could be expensive in construction and memory heavy. So per-instance type construction could yield perf/memory issues. 

I didn't found any fast an elegant solution for this. One idea was to add a reference from `ViewType` to `Spec` (as soon as a type could be Spec-dependent). In that case we could derive the type. But I would suggest to address this in further `Spec` work

Reply:
Ok makes sense.